src/Orleans/Serialization/SerializationManager.cs

```diff
@@ -390,31 +395,79 @@ public static void Register(Type t, DeepCopier cop, Serializer ser, Deserializer
                 {
                     if (cop != null)
                     {
-                        lock (copiers)
+                        if (_setupDone)
                         {
-                            DeepCopier current;
-                            if (forceOverride || !copiers.TryGetValue(t.TypeHandle, out current) || (current == null))
+                            lock (copiers)
```

Comment:
I believe the double locking was previously being used to avoid issues when reading, but since you removed the locks when reading (as the dictionaries are now effectively immutable) you probably don't need the seconds layer and more granular locking.
Please verify, as you are looking at the code more closely at this point.